### ISSUE TITLE
Feature Request: Provide ability to mention requesting account name in bot's reply

### ISSUE THREAD

**Issue by toushin-taishi** [8go/matrix-eno-bot#18]
Hello @8go,

In a busy/chatty room, it would be convenient for the people in the room if the  BOT will include the requesting account name in its reply. 

I hope you would consider this request in the future release of matrix-eno-bot.

Thank you for your hard work!

**Comment by 8go** [comment]
This is a very meaningful request @toushin-taishi  Thank you!

With this feature request, the bot could respond "Good morning User123" to user `User123`. Many use cases for that.

Right now I have limited time, but maybe somebody else gets animated and wants to implement this in a Pull Request. All contributions are welcome! :heart:

**Comment by troglodyne** [comment]
Took a stab at it myself tonight, let me know if you like the approach or whether you've got a better idea. If so I'd be cool with reworking it.

**Comment by toushin-taishi** [comment]
Hi @troglodyne,

Thank you for looking into this feature request.
I tested your changes and the following are my test results:
1. If the script does not have an argument (example: hello.sh), it properly returns @sender name.
2. If the script has an argument, it does not return the @sender name because the argument replaces the $1 variable.

**Comment by toushin-taishi** [comment]
Hi @troglodyne,
Read your reply at the Pull request tab and I understood the use/case scenario that you meant.
Is it possible for the BOT to always return @sender regardless if the script has an argument as long as its option is enabled in commands.yaml? Thanks again for looking into this.

**Comment by troglodyne** [comment]
I would need to add this, it wouldn't be difficult to do. Perhaps a "always pass in default args first" option to commands.yaml?

Will play around with it later today

**Comment by troglodyne** [comment]
OK, I think I got it figured:
https://github.com/troglodyne/matrix-eno-bot/commit/8df447ee1d4a62e68368bc8a42a4145a0a369dfa

Let me know if this doesn't address that need well

**Comment by toushin-taishi** [comment]
@troglodyne 
Hi. Sorry, it took me some time to test your changes.
Is my understanding correct that I have to remove the default_args and replace it with always_arg in commands.yaml?
Afterwhich, the following is what's happening:
1. The @sender is correctly being passed as $1 with or without argument
2. The command's argument is passed as $2. However, subsequent commands append the argument to the previous command.
3. Restarting the BOT resets the argument.

example with a custom script named "timein.sh"
1st command: timein 000000

> 2021-05-18 16:47:04,225 | bot_commands [DEBUG] OS command "timein.sh" with args: "['@toushin:foo.org', '000000']"

2nd command: timein 000111

> 2021-05-18 16:49:00,546 | bot_commands [DEBUG] OS command "timein.sh" with args: "['@toushin:foo.org', '000000', '000111']"

It is keeping the argument of the previous command as an argument to the new command given.


Kind regards

**Comment by troglodyne** [comment]
Thanks for the report. I'll have some time later in the evening today to look at fixing the oddity seen in 2.

**Comment by 8go** [comment]
OK, to start out with a big THANKS and :clap: to both of you, @troglodyne and @toushin-taishi 

I had a look at the code proposed by @troglodyne today and I would like to share my opinions with you to see what you think and to discuss alternatives and find the best (or at least) a good solution.

We need a generic solution that 
- will work with user-written scripts (e.g. where user has control over the arguments), but it must also work with system scripts (where user has no control over args), 
- is as simple as possible

In my humble opinion, the current solution implemented by @troglodyne is too complex, too many different cases and hence noobie users might get confused. 

How about the following, alternative approach: we set an environment variable, e.g. ENO_SENDER. We set the variable correctly with the sender of the incoming message. And in the file bot_commands.py

https://github.com/8go/matrix-eno-bot/blob/master/bot_commands.py#L181-L186

we add the `env` argument. See https://docs.python.org/3/library/subprocess.html#subprocess.Popen

See example: 
```
Environment example:

os.spawnlpe(os.P_NOWAIT, "/bin/mycmd", "mycmd", "myarg", env)
==>
Popen(["/bin/mycmd", "myarg"], env={"PATH": "/usr/bin"})

```

We add the ENO_SENDER with value to the `env` argument of `popen`. 


To do this we have to add the argument `sender` to this function signature:

https://github.com/8go/matrix-eno-bot/blob/master/bot_commands.py#L146-L154

Most scripts will ignore it and not know anything about $ENO_SENDER. But specific custom-scripts can check if this env variable `ENO_SENDER` is set inside the script and act and use it accordingly. 

In short, `hello.sh`, could first double check to be sure `ENO_SENDER` is set, if set it uses sender name in reply. If not set, it sends the std generic reply without name. 

This way we do not complicate the script arguments, existing scripts just work fine (never check or use ENO_SENDER) and new, custom scripts have to option to use or not use it. 

I think it is less code, less to explain to the user, easier to understand. 

What do you guys think about my idea? 

@troglodyne Do you see how this can be implemented? Do you also feel that this is simpler?

**Comment by troglodyne** [comment]
I like it. Shouldn't be hard to implement. Thanks for the suggestion!

**Comment by 8go** [comment]
I am glad to hear that you like it. 

This alternative is also simpler to extend in the future in case we want to pass even more info into the scripts, like timestamps or whatever. 

Do you have any questions regarding how to do it?

**Comment by troglodyne** [comment]
Nah, no questions about it. Setting env before shelling out is about as easy to do in python as it is in perl. When I read your suggestion I was like "dang, why didn't I think about that, that would have been way easier"

Have done a lot of "IPC via ENV" myself in the past, and in this case it makes perfect sense.

**Comment by 8go** [comment]
Excellent, fantastic ! All power to you !

**Comment by troglodyne** [comment]
I think I got it done, gonna test it on my own homeserver now

Much smaller changeset, hoping it works out good :)

**Comment by troglodyne** [comment]
Testing checking out good now, see PR for more details

Let me know what y'all think :)

**Comment by 8go** [comment]
I like it @troglodyne 

A lot less code, which I like too, :clap: 

Simpler is better. Very nice, well done!

Please read the PR comments for more comments. PR #21

**Comment by toushin-taishi** [comment]
@troglodyne and @8go,

Thanks so much for including this feature! 
I test and this works as advertise and as expected.

Again, kudos to both of you and my heartfelt appreciation for including this feature in this wonderful project.

**Comment by toushin-taishi** [comment]
> This alternative is also simpler to extend in the future in case we want to pass even more info into the scripts, like timestamps

Message timestamp is a great idea! This will allow scripts to capture the actual date/time of the message. 
Should I open a feature request for this?

**Comment by 8go** [comment]
Yes, please open a new Issue as a "Feature Request" and maybe cross reference this issue #18
